runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -927,22 +941,53 @@ public void setTimer(
     /** @deprecated use {@link #setTimer(StateNamespace, String, Instant, TimeDomain)}. */
     @Deprecated
     @Override
-    public void setTimer(TimerData timerKey) {
-      long time = timerKey.getTimestamp().getMillis();
-      switch (timerKey.getDomain()) {
+    public void setTimer(TimerData timer) {
+      try {
+        getKeyedStateBackend().setCurrentKey(getCurrentKey());
+        String uniqueTimerId = getUniqueTimerId(timer);
```

Comment:
The logic here also needs explanation. We first remove the pending timer but then it is being put back in `registerTimer`?

Reply:
It's because a new timer with timerId X resets the currently pending timer X in the same context. So we have to look up and remove the current timer and then set the new timer which involves adding it back to the "pending timers" map.